File: trunk/Crystal/CrystalUI/MainForm.cpp
#include "MainForm.h"

#include "SimulationSettingForm.h"
#include "ObjectSettingForm.h"
#include "ParticleObserveForm.h"
#include "BoundarySettingForm.h"
#include "GraphicsSettingForm.h"
#include "ExportSettingForm.h"

using namespace Crystal::Command;
using namespace Crystal::UI;

using namespace System;
using namespace System::ComponentModel;
using namespace System::Collections;
using namespace System::Windows::Forms;
using namespace System::Data;
using namespace System::Drawing;
using namespace System::Diagnostics;
using namespace System::IO;

MainForm::MainForm()
{
	InitializeComponent();

	pictureBox1->Image = gcnew Bitmap( pictureBox1->Width, pictureBox1->Height );
	command = gcnew MainCommand( pictureBox1 );
}

MainForm::~MainForm()
{
	if (components) {
		delete components;
	}
}

System::Void MainForm::MainForm_Load(System::Object^  sender, System::EventArgs^  e)
{
}

System::Void MainForm::buttonNextStep_Click(System::Object ^sender, System::EventArgs ^e) {
	command->proceedSimulation();
	command->displayInformation( listBoxInformation );
}

System::Void MainForm::timerSimulation_Tick(System::Object^  sender, System::EventArgs^  e)
{
	command->proceedSimulation();
	command->displayInformation( listBoxInformation );
}

System::Void MainForm::buttonStart_Click(System::Object ^sender, System::EventArgs ^e) {
	timerSimulation->Enabled = !timerSimulation->Enabled;
}

System::Void MainForm::buttonSimulationRefresh_Click(System::Object ^sender, System::EventArgs ^e) {
	command->refreshSimulation();
	command->displayInformation( listBoxInformation );
}

System::Void MainForm::simulationSettingToolStripMenuItem_Click(System::Object^  sender, System::EventArgs^  e)
{
	SimulationSettingForm^ form = gcnew SimulationSettingForm( command->getSimulationSettingCommand() );
	form->Show();
}

System::Void MainForm::sToolStripMenuItem_Click(System::Object^  sender, System::EventArgs^  e)
{
	SimulationSettingForm^ form = gcnew SimulationSettingForm( command->getSimulationSettingCommand() );
	form->Show();
}

System::Void MainForm::objectSettingToolStripMenuItem_Click(System::Object ^sender, System::EventArgs ^e)
{
	ObjectSettingForm^ form = gcnew ObjectSettingForm( command->getObjectSettingCommand() );
	form->Show();
}

System::Void MainForm::toolStripMenuItemOpen_Click(System::Object^  sender, System::EventArgs^  e)
{
	command->getFileIOCommand()->fileOpen();
}

System::Void MainForm::toolStripMenuItemSaveAs_Click(System::Object^  sender, System::EventArgs^  e)
{
	command->getFileIOCommand()->fileSave();
}

System::Void MainForm::toolStripMenuItemSave_Click(System::Object^  sender, System::EventArgs^  e)
{
	command->getFileIOCommand()->fileSaveOverride();
}

System::Void MainForm::toolStripMenuItemClose_Click(System::Object^  sender, System::EventArgs^  e) {
	Close();
}

System::Void MainForm::MainForm_DragEnter(System::Object^  sender, System::Windows::Forms::DragEventArgs^  e){
	if( e->Data->GetDataPresent( System::Windows::Forms::DataFormats::FileDrop ) ) {
		e->Effect = System::Windows::Forms::DragDropEffects::All;
	}
}

System::Void MainForm::MainForm_DragDrop(System::Object^ sender, System::Windows::Forms::DragEventArgs^ e ) {
	array<String^>^ files = dynamic_cast<array<String^>^>(
		e->Data->GetData( DataFormats::FileDrop ) );
	String^ path = files[0];

	if( !command->getFileIOCommand()->fileOpen( path ) ) {
		MessageBox::Show("Unexpected File Format");
	}
	System::String^ fileName = path;
}

System::Void MainForm::particleInfoToolStripMenuItem_Click(System::Object^  sender, System::EventArgs^  e) {
	ParticleObserveForm^ form = gcnew ParticleObserveForm( command->getParticleInfoCommand() );
	form->Show();
	form->onDisplay();
}

System::Void MainForm::objectInfoOToolStripMenuItem_Click(System::Object^  sender, System::EventArgs^  e) {
}

System::Void MainForm::boundarySettingBToolStripMenuItem_Click(System::Object^  sender, System::EventArgs^  e) {
	BoundarySettingForm^ form = gcnew BoundarySettingForm( command->getBoundarySettingCommand() );
	form->Show();
}

System::Void MainForm::graphicsSettingGToolStripMenuItem_Click(System::Object^  sender, System::EventArgs^  e) {
	GraphicsSettingForm^ form = gcnew GraphicsSettingForm( command );
	form->Show();
}

System::Void MainForm::saveImagesIToolStripMenuItem_Click(System::Object ^sender, System::EventArgs ^e) {
	ExportSettingForm^ form = gcnew ExportSettingForm( command->getExportSettingCommand() );
	form->Show();
}

System::Void MainForm::exportParticleEToolStripMenuItem_Click(System::Object^  sender, System::EventArgs^  e){
	ExportSettingForm^ form = gcnew ExportSettingForm( command->getParticleExportSettingCommand() );
	form->Show();
}

System::Void MainForm::simulationSettingSToolStripMenuItem_Click(System::Object ^sender, System::EventArgs ^e){
	simulationSettingToolStripMenuItem_Click(sender, e );
}

System::Void MainForm::MainForm_KeyDown(System::Object^  sender, System::Windows::Forms::KeyEventArgs^  e){
	switch( e->KeyCode ) {
	case System::Windows::Forms::Keys::D :
		command->zoom(0.1);
		break;
	case System::Windows::Forms::Keys::Z :
		command->zoom(-0.1);
		break;
	}
}

System::Void MainForm::MainForm_MouseWheel(System::Object^ , System::Windows::Forms::MouseEventArgs^ e){
	const double zoom = -e->Delta / 240.0;
	command->zoom(zoom);
}

System::Void MainForm::pictureBox1_MouseMove(System::Object^  sender, System::Windows::Forms::MouseEventArgs^  e) {
	if( isLeftDown && ( e->Button == System::Windows::Forms::MouseButtons::Left ) ) {
		double diffX = -(previousPoint.X - e->Location.X) / 1000.0;
		double diffY = -(previousPoint.Y - e->Location.Y) / 1000.0;
		System::Diagnostics::Debug::WriteLine( "diffX = " + diffX + "diffY = " + diffY );
		command->move( diffX, diffY );
		previousPoint = e->Location;
	}
	else if( isRightDown && ( e->Button == System::Windows::Forms::MouseButtons::Right ) ) {
		int angleY = -(previousPoint.X - e->Location.X);
		int angleX = -(previousPoint.Y - e->Location.Y);
		command->rotateX(angleX);
		command->rotateY(angleY);
		previousPoint = e->Location;
	}
}

System::Void MainForm::pictureBox1_MouseDown(System::Object^  sender, System::Windows::Forms::MouseEventArgs^  e) {
	if( e->Button == System::Windows::Forms::MouseButtons::Left ) {
		previousPoint = e->Location;
		isLeftDown = true;
	}
	if( e->Button == System::Windows::Forms::MouseButtons::Right ) {
		previousPoint = e->Location;
		isRightDown = true;
	}
}

System::Void MainForm::pictureBox1_MouseUp(System::Object^  sender, System::Windows::Forms::MouseEventArgs^  e){
	isLeftDown = false;
	isRightDown = false;
}

System::Void MainForm::buttonViewReset_Click(System::Object^  sender, System::EventArgs^  e) {
	command->viewReset();
}

System::Void MainForm::buttonParticleInfo_Click(System::Object^  sender, System::EventArgs^  e) {
	ParticleObserveForm^ form = gcnew ParticleObserveForm( command->getParticleInfoCommand() );
	form->Show();
	form->onDisplay();
}

File: Cryphous/Cryphous/MainForm.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using Crystal.Command;
using System.IO;

namespace Cryphous
{
    public partial class MainForm : Form
    {
        private BoundarySettingCommand boundarySettingCommand;
        private ObjectSettingCommand objectSettingCommand;
        private ProfileInfoCommand profileInfoCommand;
        private ParticleInfoCommand particleInfoCommand;
        private SimulationSettingCommand simulationSettingCommand;
        private MainCommand mainCommand;
        private Point previousPoint;
	    private bool isLeftDown;
        private bool isRightDown;

        private PhysicsObjectSettingDialog osDialog;
        private GraphicsSettingForm gsDialog;

        private bool isStandAlone;
        List<List<float[]>> simulatedPositions;
        List<List<float[]>> simulatedNormals;

        private int maxParticles;

        public List<List<float[]>> SimulatedPositions
        {
            get{ return simulatedPositions; }
        }

        public List<List<float[]>> SimulatedNormals
        {
            get { return simulatedNormals; }
        }
        
        public MainForm(bool isStandAlone, int maxParticles)
        {
            InitializeComponent();
            this.pictureBox1.MouseWheel += new System.Windows.Forms.MouseEventHandler(this.pictureBox1_MouseWheel);
            this.isStandAlone = isStandAlone;
            this.maxParticles = maxParticles;
        }

        private void MainForm_Load(object sender, EventArgs e)
        {
            ApplicationSettings.get();
            boundarySettingCommand = new BoundarySettingCommand();
            objectSettingCommand = new ObjectSettingCommand();
            profileInfoCommand = new ProfileInfoCommand();
            particleInfoCommand = new ParticleInfoCommand();
            simulationSettingCommand = new SimulationSettingCommand();
            pictureBox1.Image = new Bitmap(pictureBox1.Width, pictureBox1.Height);
            mainCommand = new MainCommand(pictureBox1, "Cryphous 1.3");

            osDialog = new PhysicsObjectSettingDialog(objectSettingCommand, boundarySettingCommand, simulationSettingCommand, maxParticles);
            gsDialog = new GraphicsSettingForm(mainCommand);
            objectSettingToolStripMenuItem_Click(sender, e);
        }

        private void objectSettingToolStripMenuItem_Click(object sender, EventArgs e)
        {
            osDialog.ShowDialog();
            if (!isStandAlone)
            {
                simulatedPositions = mainCommand.getSimulationCommand().getManagedPositions();
                simulatedNormals = mainCommand.getSimulationCommand().getManagedNormals();
            }
        }

        private void buttonStart_Click(object sender, EventArgs e)
        {
            timerSimulation.Enabled = !timerSimulation.Enabled;
        }

        private void buttonRefresh_Click(object sender, EventArgs e)
        {
            mainCommand.refreshSimulation();
            mainCommand.displayInformation(listBoxInformation);
        }

        private void toolStripMenuItemClose_Click(object sender, EventArgs e)
        {
            Close();
        }

        private void buttonNextStep_Click(object sender, EventArgs e)
        {
            proceed();
        }

        private void timerSimulation_Tick(object sender, EventArgs e)
        {
            proceed();
        }

        public void proceed()
        {
            mainCommand.proceedSimulation();
            mainCommand.displayInformation(listBoxInformation);
            if (!isStandAlone)
            {
                simulatedPositions = mainCommand.getSimulationCommand().getManagedPositions();
                simulatedNormals = mainCommand.getSimulationCommand().getManagedNormals();
            }
        }

        private void MainForm_KeyDown(object sender, KeyEventArgs e)
        {
            switch( e.KeyCode ) {
	            case Keys.D :
            		mainCommand.zoom(1.0F);
		            break;
            	case Keys.Z :
	               	mainCommand.zoom(-1.0F);
		        break;
	        }
        }

        private void buttonViewReset_Click(object sender, EventArgs e)
        {
            mainCommand.viewReset();
        }

        private void pictureBox1_MouseDown(object sender, MouseEventArgs e)
        {
            if (e.Button == MouseButtons.Left)
            {
                previousPoint = e.Location;
                isLeftDown = true;
            }
            if (e.Button == MouseButtons.Right)
            {
                previousPoint = e.Location;
                isRightDown = true;
            }
        }

        private void pictureBox1_MouseMove(object sender, MouseEventArgs e)
        {
            if (isLeftDown && (e.Button == MouseButtons.Left))
            {
                float diffX = -(previousPoint.X - e.Location.X) / 1000.0F;
                float diffY = -(previousPoint.Y - e.Location.Y) / 1000.0F;
                mainCommand.move(diffX, diffY);
                previousPoint = e.Location;
            }
            else if (isRightDown && (e.Button == MouseButtons.Right))
            {
                int angleY = -(previousPoint.X - e.Location.X);
                int angleX = -(previousPoint.Y - e.Location.Y);
                mainCommand.rotateX(angleX);
                mainCommand.rotateY(angleY);
                previousPoint = e.Location;
            }
        }

        private void pictureBox1_MouseUp(object sender, MouseEventArgs e)
        {
            isLeftDown = false;
            isRightDown = false;
        }

        private void pictureBox1_MouseWheel(object sender, MouseEventArgs e)
        {
            float zoom = -e.Delta / 24.0F;
            mainCommand.zoom(zoom);
        }

        private void pictureBox1_Click(object sender, EventArgs e)
        {
            pictureBox1.Focus();
        }

        private void graphicsSettingGToolStripMenuItem_Click(object sender, EventArgs e)
        {
            gsDialog.Show();
        }

        private void buttonParticleInfo_Click(object sender, EventArgs e)
        {
            ParticleObserveDialog poDialog = new ParticleObserveDialog(particleInfoCommand);
            poDialog.ShowDialog();
        }
    }
}
